content/blog/2023-08-03-parsing-ssh-config.markdown

```diff
@@ -0,0 +1,229 @@
+system wide settings. Most of the time this isn't an issue because of the
+first-come-first-served rule with config matches (rule 2). But if the system
+config includes a `Match final`, it will trigger the entire config to be
+re-parsed, including the user section. And it so happens that, at least on
+Fedora with the `openssh-clients` package installed, the system config does
+contain a `Match final`.
+
+But wait, there's more! If we want to specify a custom SSH config file, then we
+can use `-F path/to/config` in the command line. But this disables loading a
+system config, so we would no longer get the proxy!
+
+To sum up, for the above config:
+
+1. `ssh example` doesn't have a proxy
+2. ...unless a system config contains `Match final`
+3. ...but invoking it as `ssh -F ~/.ssh/config example` definitely won't have
+   the proxy
+4. ...but if a subprocess invokes `ssh example` while trying to resolve another
+   host, it'll probably not add the `-F ~/.ssh/config`, so we might get the
+   proxy again (in the child process).
+
+Wait, how did that last one slip in? Well, unlike environment variables, it's a
+lot harder for processes to propagate command line flags correctly. If resolving
+the config involves running a script that itself tries to run SSH, chances are
+the `-F` flag won't be propagated and you'll see some weird behaviour.
+
+I swear that's all for now, you've probably learned more about SSH configs than
+you will ever need to care about.
+
+
+## Back to VSCodium
+
+Alright, armed now with this knowledge on SSH config parsing, we can work out
+what's going on with the extension. It ends up being a simple issue: it doesn't
+apply rule (3), so all `Host` matches are done against the original host name.
+
+In my case, there are several machines behind the proxy, but they all share a
+common suffix, so I had a `Host *.ozlabs.ibm.com` rule to apply the proxy. I
+also use aliases to refer to the machines without the `.ozlabs.ibm.com` suffix,
+so failing to follow rule (3) lead to the situation where the extension didn't
+think there was a proxy.
+
+However, even if this were to be fixed, it still doesn't respect rule (4), or
+most complex match logic in general. If the hostname bug is fixed then my setup
+would work, but it's less than ideal to keep playing whack-a-mole with parsing
+bugs. It would be a lot easier if there was a way to just ask SSH for the config
+that a given host name resolves to.
+
+Enter `ssh -G`. The `-G` flag asks SSH to dump the complete resolved config,
+without actually opening the connection (it may execute arbitrary code while
+resolving the config however!). So to fix the extension once and for all, we
+could swap the manual parser to just invoking `ssh -G example`, and parsing the
+output as the final config. No `Host` or `Match` or `HostName` or `Match final`
+quirks to worry about.
+
+Sure enough, if we replace the config backend with this 'native' resolver, we
```

Comment:
are you gonna upstream this

Reply:
I've made a PR. And I'm working on fixing the parser too

Reply:
cool, throw in a link